types/jest/index.d.ts

```diff
@@ -693,24 +693,35 @@ declare namespace jest {
         getState(): MatcherState & Record<string, any>;
     }
 
-    type JestMatchers<T> = JestMatchersShape<Matchers<void, T>, Matchers<Promise<void>, T>>;
+    type PromiseOf<T> = T extends Promise<infer S> ? S : T;
+    type JestMatchers<T> = JestMatchersShape<Matchers<void, T>, Matchers<Promise<void>, PromiseOf<T>>, Matchers<Promise<void>, any>>;
 
-    type JestMatchersShape<TNonPromise extends {} = {}, TPromise extends {} = {}> = {
+    type JestMatchersShape<TNonPromise extends {} = {}, TResolve extends {} = {}, TReject extends {} = {}> = {
         /**
          * Use resolves to unwrap the value of a fulfilled promise so any other
          * matcher can be chained. If the promise is rejected the assertion fails.
          */
-        resolves: AndNot<TPromise>,
+        resolves: AndNot<TResolve>,
         /**
          * Unwraps the reason of a rejected promise so any other matcher can be chained.
          * If the promise is fulfilled the assertion fails.
          */
-        rejects: AndNot<TPromise>
+        rejects: AndNot<TReject>
     } & AndNot<TNonPromise>;
     type AndNot<T> = T & {
         not: T
     };
 
+    type JasmineMatcher =
+        | ReturnType<typeof jasmine.any>
+        | ReturnType<typeof jasmine.arrayContaining>
+        | ReturnType<typeof jasmine.objectContaining>;
+
+    type SafeParameters<T> = T extends (...args: any) => any ? Parameters<T> : any[];
+    type SafeReturn<T> = T extends (...args: any) => any ? ReturnType<T> : any[];
```

Comment:
I'm confused by these names. What makes them "safe"?

Reply:
`Parameters<T>` requires `T` to extend `(...args: any) => any`. I'm happy to rename this to a better name, this is the best I could come up with.

Reply:
SafeReturn should return any instead of any[]

Reply:
`MaybeParameters` and `MaybeReturn`?